svm/src/nonce_info.rs

```diff
@@ -14,10 +23,39 @@ pub struct NoncePartial {
     account: AccountSharedData,
 }
 
+#[derive(Error, Debug, PartialEq)]
+pub enum AdvanceNonceError {
+    #[error("Invalid account")]
+    Invalid,
+    #[error("Uninitialized nonce")]
+    Uninitialized,
+}
+
 impl NoncePartial {
     pub fn new(address: Pubkey, account: AccountSharedData) -> Self {
         Self { address, account }
     }
+
+    // Advance the stored blockhash to prevent fee theft by someone
+    // replaying nonce transactions that have failed with an
+    // `InstructionError`.
+    pub fn try_advance_account(
+        &mut self,
+        durable_nonce: DurableNonce,
+        lamports_per_signature: u64,
+    ) -> Result<(), AdvanceNonceError> {
+        let nonce_versions = StateMut::<NonceVersions>::state(&self.account)
+            .map_err(|_| AdvanceNonceError::Invalid)?;
+        if let NonceState::Initialized(ref data) = nonce_versions.state() {
+            let nonce_state =
+                NonceState::new_initialized(&data.authority, durable_nonce, lamports_per_signature);
+            let nonce_versions = NonceVersions::new(nonce_state);
+            self.account.set_state(&nonce_versions).unwrap();
+            Ok(())
+        } else {
+            Err(AdvanceNonceError::Uninitialized)
```

Comment:
(note to self: statically_, this `Err(AdvanceNonceError::Uninitialized)` could be `.unwrap()`-ed after calling `.try_advance_account()`. This is a new potential source of `.unwrap()` compared to the prev code. However, it seems not possible for `nonce_versions` to be `NonceState::Legacy` here. That's because `Versions::verify_recent_blockhash()` is run for all relevant code-path, which sanitizes nonces correctly. Thus, it's safe.)

Reply:
Yup, but just to point out that instead of saying "it seems not possible for `nonce_versions` to be `NonceState::Legacy` here, I think you meant to say `NonceState::Uninitialized`.

Reply:
oops, thanks for correction. edited my comment.